File: Data Structures & Algorithms/permutations/submission-0.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> sol = new ArrayList<>();
        bt(nums, new ArrayList<>(), new HashSet<>(), sol);
        return sol;
    }

    void bt(int[] nums, List<Integer> curr, Set<Integer> visited, List<List<Integer>> sol) {
        if (curr.size() == nums.length) {
            sol.add(new ArrayList<>(curr));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited.contains(i)) continue;
            curr.add(nums[i]);
            visited.add(i);
            bt(nums, curr, visited, sol);
            visited.remove(i);
            curr.remove(curr.size()-1);
        }
    }
}

File: Data Structures & Algorithms/target-sum/submission-0.go
func findTargetSumWays(nums []int, target int) int {
	n := len(nums)

    var bt func (i, curr int) int
	bt = func (i, curr int) int {
		if i == n {
			if curr == target {
				return 1
			}
			return 0
		}

		return bt(i+1, curr + nums[i]) + bt(i+1, curr - nums[i])
	}

	return bt(0, 0)
}
